homeassistant/components/system_bridge/config_flow.py

```diff
@@ -37,75 +36,6 @@
 )
 
 
-async def _validate_input(
-    hass: HomeAssistant,
-    data: dict[str, Any],
-) -> dict[str, str]:
-    """Validate the user input allows us to connect.
-
-    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
-    """
-    host = data[CONF_HOST]
-
-    websocket_client = WebSocketClient(
-        host,
-        data[CONF_PORT],
-        data[CONF_API_KEY],
-    )
-    try:
-        async with asyncio.timeout(15):
-            await websocket_client.connect(session=async_get_clientsession(hass))
-            hass.async_create_task(websocket_client.listen())
-            response = await websocket_client.get_data(GetData(modules=["system"]))
-            _LOGGER.debug("Got response: %s", response.json())
-            if response.data is None or not isinstance(response.data, System):
-                raise CannotConnect("No data received")
-            system: System = response.data
-    except AuthenticationException as exception:
-        _LOGGER.warning(
-            "Authentication error when connecting to %s: %s", data[CONF_HOST], exception
-        )
-        raise InvalidAuth from exception
-    except (
-        ConnectionClosedException,
-        ConnectionErrorException,
-    ) as exception:
-        _LOGGER.warning(
-            "Connection error when connecting to %s: %s", data[CONF_HOST], exception
-        )
-        raise CannotConnect from exception
-    except TimeoutError as exception:
-        _LOGGER.warning("Timed out connecting to %s: %s", data[CONF_HOST], exception)
-        raise CannotConnect from exception
-    except ValueError as exception:
-        raise CannotConnect from exception
-
-    _LOGGER.debug("Got System data: %s", system.json())
-
-    return {"hostname": host, "uuid": system.uuid}
-
-
-async def _async_get_info(
-    hass: HomeAssistant,
-    user_input: dict[str, Any],
-) -> tuple[dict[str, str], dict[str, str] | None]:
-    errors = {}
-
-    try:
-        info = await _validate_input(hass, user_input)
-    except CannotConnect:
-        errors["base"] = "cannot_connect"
-    except InvalidAuth:
-        errors["base"] = "invalid_auth"
-    except Exception:  # pylint: disable=broad-except
-        _LOGGER.exception("Unexpected exception")
-        errors["base"] = "unknown"
-    else:
-        return errors, info
-
-    return errors, None
-
-
 class SystemBridgeConfigFlow(
```

Comment:
Looks like a bug to set `_system_data` as an instance attribute. I think that should just be a local variable, since we want to be able to check the connection multiple independent times.

https://github.com/home-assistant/core/blob/2e65952eae33e336ed9be13864711aa3201c8faf/homeassistant/components/system_bridge/config_flow.py#L141-L163

This also means there's not much reason to have those methods, `_async_get_info` and `_validate_input`, as methods and they can be moved out from the class to functions. The only remaining instance attribute they need is `hass`, which can be passed as a parameter.

Reply:
These were moved into the class previously to avoid using `nonlocal system_data`:

https://github.com/home-assistant/core/pull/107957#discussion_r1484139991

Reply:
I'm not sure how to go about this without using `nonlocal` or `global` for system_data?

Reply:
You could update a mutable object like a dict, stored in a local variable.